packages/databus/src/index.ts

```diff
@@ -0,0 +1,133 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+import { DisposableDelegate, IDisposable } from '@phosphor/disposable';
+
+import { ArrayExt } from '@phosphor/algorithm';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+/**
+ * An interface for an abstract dataset, with a mimetype, data, and metadata.
+ *
+ * #### Notes
+ * This interface is similar conceptually to that of nbformat.IMimeBundle,
+ * but with only a single mimetype and different structure. We expect there
+ * to be utilities to convert between this formats.
+ */
+export interface IDataset {
+  /**
+   * The string mimetype for the dataset.
+   */
+  mimeType: string;
+
+  /***
+   * The data of the dataset.
+   *
+   * #### Notes
+   * The notion of data in the databus is completely abstract, and typed as an
+   * any. Publishers and consumers of a given mimetype will need to agree
+   * what the actual data is and cast it appropriately before use.
+   */
+  data: any;
+
+  /**
+   * The metadata associate with the dataset.
+   *
+   * #### Notes
+   * Like the data itself, the databus notion of metadata is entirely abstract.
+   * We are currently exploring different metadata schemes, so this may change
+   * in the future.
+   */
+  metadata: any;
+}
+
+/**
+ * A databus object, for managing datasets in JupyterLab.
+ */
+export class DataBus {
+  /**
+   * Construct a new databus.
+   */
+  constructor() {}
+
+  /**
+   * Publish a dataset to the databus.
+   *
+   * @param dataset - the `IDataset` to publish to the databus.
+   *
+   * @returns A disposable which will remove the dataset.
+   *
+   * @throws An error if the given dataset is already published.
+   */
+  publish(dataset: IDataset): IDisposable {
+    if (this._datasets.find(i => i === dataset) === undefined) {
+      throw new Error(`Dataset already published`);
+    }
+
+    var index = this._datasets.push(dataset) - 1;
+
+    this._datasetsChanged.emit({ dataset, type: 'added' });
+
+    return new DisposableDelegate(() => {
+      ArrayExt.removeAt(this._datasets, index);
```

Comment:
I think the index might not necessarily be valid when it is removed. For example, let's say A is added, then B is added. So at adding time, B is at index 1. Then A is removed. Now, B is at index 0. But, the index here will still be 1.

Reply:
Good catch, push fix.